File: Math/TypeScript/_0202_HappyNumber.ts
/**
 * Iteration
 * Time Complexity: BigO(N^2)
 * Space Complexity: BigO(1)
 */
function isHappyV1(n: number): boolean {
  let num = n;
  while (num > 6) {
    let nl = num.toString().split("");
    num = nl.reduce((acc, val) => {
      return (acc += +val * +val);
    }, 0);
  }
  return num == 1 ? true : false;
}

/**
 * Iteration
 * Time Complexity: BigO(N)
 * Space Complexity: BigO(1)
 */
function isHappyV2(n: number): boolean {
  let fast = n,
    slow = n;

  do {
    fast = sumOfSquares(sumOfSquares(fast));
    slow = sumOfSquares(slow);
  } while (fast !== slow);

  return slow === 1;
}

function sumOfSquares(n: number): number {
  let num = 0;
  while (n !== 0) {
    const dig = n % 10;
    num += dig * dig;
    n = Math.floor(n / 10);
  }
  return num;
}

console.log(isHappyV2(19));
console.log(isHappyV2(2));
console.log(isHappyV2(3));
console.log(isHappyV2(7));

File: Math/Python/_0202_HappyNumber.py
class Solution:
    # Iteration
    # Time Complexity: BigO(N)
    # Space Complexity: BigO(N)
    def isHappy(self, n: int) -> bool:
        def sumOfSquares(n: int) -> int:
            num = 0
            while n:
                num += (n % 10) ** 2
                n = n // 10
            return num

        lst = set()
        while True:
            if n == 1:
                return True
            elif n in lst:
                return False
            lst.add(n)
            n = sumOfSquares(n)


ins = Solution()

print(ins.isHappy(19))
print(ins.isHappy(2))
print(ins.isHappy(3))
print(ins.isHappy(7))
